server/connectors/bookmarks.js

```diff
@@ -4,41 +4,110 @@
+ *
+ * At some point, we'd like to have our own DB of bookmarks, and at that point
+ * should add infinite-scrolling to the client so the client can request only
+ * those results it actually needs to fill the page.
  */
 
 
 const fetch = require('node-fetch');
 
-const RAINDROP_REST_URL = `https://raindrop.io/api/raindrops/2021037`;
+const PRESTERITY_BOOKMARK_COLLECTION_ID = 2021037;
+const RAINDROP_REST_URL = `https://raindrop.io/api/raindrops/${PRESTERITY_BOOKMARK_COLLECTION_ID}`;
 const MAX_BOOKMARKS_PER_PAGE = 40; // Limit imposed by Raindrop.
 
 
+/*
+ * Return a promise for the complete set of bookmarks on the given topic.
+ * The topic is identified with a string: "Trump Cabinet".
+ *
+ * The bookmarks are returned sorted by bookmark title. Since we start bookmark
+ * titles with a date in YYYY.MM.DD format, the bookmarks will also be sorted
+ * by date.
+ */
 function bookmarksForTopic(topic) {
-  const url = `${RAINDROP_REST_URL}?search=[{"key":"tag","val":"${topic}"}]&perpage=${MAX_BOOKMARKS_PER_PAGE}`;
-  console.log(`Bookmarks: ${url}`);
-  return fetch(url)
-  .then(response => response.json())
-  .then(json => {
-    const sorted = json.items.sort(compareItems);
+
+  let bookmarks;
+
+  // Fetch the initial results: page 0.
+  return getResultsForPage(topic, 0)
+  .then(resultPage0 => {
+
+    // Save the page 0 bookmarks.
+    bookmarks = resultPage0.items;
+
+    // The results include a `count` property that gives the total number of
+    // bookmarks. Use this to calculate how many more pages we need fetch.
+    const bookmarkCount = resultPage0.count;
+    const pageCount = Math.ceil(bookmarkCount / MAX_BOOKMARKS_PER_PAGE);
+
+    // Now get pages 1..pageCount.
+    return getResultsForPages(topic, pageCount);
+  })
+  .then(additionalResultPages => {
+
+    // Combine bookmarks from additional pages (if any) with set from page 0.
+    additionalResultPages.forEach(resultPage => {
+      bookmarks = bookmarks.concat(resultPage.items);
+    });
+
+    // Raindrop's sorting facilities are limited, so we sort ourselves.
+    const sorted = bookmarks.sort(compareBookmarks);
+
     return sorted;
   });
 }
 
-// Sort two items by title.
+// Sort two bookmarks by title.
 // Since the title should start with a date like 2017.01.13, sorting by title
 // should produce a chronological sort.
-function compareItems(a, b) {
-  const aValue = a.title ? a.title.toLowerCase() : '';
-  const bValue = b.title ? b.title.toLowerCase() : '';
-  if (aValue < bValue) {
+function compareBookmarks(bookmark1, bookmark2) {
+  if (bookmark1.title < bookmark2.title) {
     return -1;
   }
-  if (aValue > bValue) {
+  if (bookmark1.title > bookmark2.title) {
     return 1;
   }
   return 0;
 }
 
+// Return a promise for the indicated page of results for the given topic.
+function getResultsForPage(topic, pageNumber) {
+  const url = `${RAINDROP_REST_URL}?search=[{"key":"tag","val":"${topic}"}]&perpage=${MAX_BOOKMARKS_PER_PAGE}&page=${pageNumber}`;
```

Comment:
It looks like the correct way to do that is `encodeURIComponent` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent), yes?

It looks like any user-entered garbage will actually cause Express to choke and parsing the page title argument _before_ the title ultimately becomes the `topic` above. If I run what's on master locally (i.e., not with this PR), and search for a garbage topic like http://localhost:8000/reference/Afghanistan%22%7D%5, then I get:

> `URIError: Failed to decode param 'Afghanistan%22%7D%5'`

It looks like that is [known behavior](http://stackoverflow.com/a/33005652/76472) with Express. The SO commenter suggest that choking on a badly-formed URL is the user's problem.

In any event, it seems good to have defense in depth. Just want to make sure I'm doing that correctly. Since this is a fairly minor point, and in the interests of time, I'm going to land this PR so that the writers — who have now hit the 40 bookmark limit — can keep going. But we can keep this thread active until I've got the topic escaping correct.